src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/TransactionValidator.cs

```diff
@@ -0,0 +1,54 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using static Hl7.Fhir.Model.Bundle;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources.Bundle
+{
+    public static class TransactionValidator
+    {
+        public static void ValidateTransaction(HashSet<string> resourceIdList, EntryComponent entry)
+        {
+            if (ValidateBundleEntry(entry))
+            {
+                string resourceId = GetResourceUrl(entry);
+
+                if (!string.IsNullOrEmpty(resourceId))
+                {
+                    if (resourceIdList.Contains(resourceId))
+                    {
+                        throw new RequestNotValidException(string.Format(Api.Resources.ResourcesMustBeUnique, entry.Request.Url));
+                    }
+
+                    resourceIdList.Add(resourceId);
+                }
+            }
+        }
+
+        private static bool ValidateBundleEntry(EntryComponent entry)
+        {
+            string requestUrl = entry.Request.Url;
+
+            // Check for duplicate resources within a bundle entry is skipped if the entry is bundle or if the request within a entry is not modifying the resource.
+            return !(entry.Resource.ResourceType == Hl7.Fhir.Model.ResourceType.Bundle
+                || entry.Request.Method == HTTPVerb.GET
+                || (entry.Request.Method == HTTPVerb.POST && requestUrl.Contains("_search", StringComparison.InvariantCulture))
+                || requestUrl.Contains("$", StringComparison.InvariantCulture));
+        }
+
+        private static string GetResourceUrl(EntryComponent component)
+        {
+            if (component.Request.Method == HTTPVerb.POST)
+            {
+                return component.FullUrl;
```

Comment:
What happens if the request is conditional create/update? In that case, the request URL contains search parameters; would that be addressed as part of work mentioned below?

Reply:
Also, what do we do in the case the URL is specified absolute/relative that points to the same resource?

Reply:
> What happens if the request is conditional create/update? In that case, the request URL contains search parameters; would that be addressed as part of work mentioned below?

I am addressing the conditional create/update scenario as a part of the next story I am working on as they are similar.

Reply:
> Also, what do we do in the case the URL is specified absolute/relative that points to the same resource?

We are returning FullURL only in POST. We are using the request.url in other cases( i.e. for PUT and DELETE). In case for POST, fullURl is mostly uuid as the resource willnot have logicalId yet. POST and PUT can be creating the same resource in case of conditional create/update when the search parameter is the same identifier. I am addressing that as a part of my story that I am currently working on. Does this answer your question?

Reply:
> We are returning FullURL only in POST.

This method is processing the incoming request, crafted by the caller, right?